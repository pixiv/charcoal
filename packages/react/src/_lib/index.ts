/**
 * 今後ポートされる予定の汎用的な関数群
 */

/**
 * Function used to assert a given code path is unreachable
 */
export function unreachable(): never
/**
 * Function used to assert a given code path is unreachable.
 * Very useful for ensuring switches are exhaustive:
 *
 * ```ts
 * switch (a.type) {
 *   case Types.A:
 *   case Types.B:
 *     break
 *   default:
 *     unreachable(a) // will cause a build error if there was
 *                    // a Types.C that was not checked
 * }
 * ```
 *
 * @param value Value to be asserted as unreachable
 */
// NOTE: Uses separate overloads, _not_ `value?: never`, to not allow `undefined` to be passed

export function unreachable(value: never): never
export function unreachable(value?: never): never {
  throw new Error(
    arguments.length === 0
      ? 'unreachable'
      : `unreachable (${JSON.stringify(value)})`
  )
}

/**
 * 複数のrefをマージする。
 *
 * forwardRefで受け取ったrefと、コンポーネント内で定義したrefを同じ要素につけたいケースなどで使う
 */
export function mergeRefs<T>(...refs: React.Ref<T>[]): React.RefCallback<T> {
  return (value) => {
    for (const ref of refs) {
      if (typeof ref === 'function') {
        ref(value)
      } else if (ref !== null) {
        // TODO: use RefObject
        // https://react.dev/blog/2024/04/25/react-19-upgrade-guide#:~:text=%3D%201%3B-,MutableRef,-is%20now%20deprecated
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    }
  }
}

export function countCodePointsInString(string: string): number {
  // [...string] とするとproduction buildで動かなくなる
  // cf. https://twitter.com/f_subal/status/1497214727511891972
  return Array.from(string).length
}
