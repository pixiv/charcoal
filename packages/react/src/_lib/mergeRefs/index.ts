import React from 'react'

/**
 * Custom implementation of mergeRefs to replace @react-aria/utils
 * Merges multiple refs into a single ref callback
 */
export function mergeRefs<T>(
  ...refs: (React.Ref<T> | undefined)[]
): React.RefCallback<T> {
  return (value: T | null) => {
    refs.forEach((ref) => {
      if (ref == null) {
        return
      }

      if (typeof ref === 'function') {
        ref(value)
      } else if (typeof ref === 'object') {
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    })
  }
}
