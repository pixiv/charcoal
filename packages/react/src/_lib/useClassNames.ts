import { useMemo } from 'react'

/**
 * Join some class names if propsClassName is defined.
 */
export function useClassNames(...classNames: (string | undefined)[]): string {
  return useMemo(() => classNames.filter((v) => v).join(' '), [classNames])
}
