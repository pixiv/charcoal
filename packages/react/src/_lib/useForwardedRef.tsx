import * as React from 'react'

export function useForwardedRef<T>(ref: React.ForwardedRef<T>): React.RefObject<T> {
  const innerRef = React.useRef<T>(null)

  React.useEffect(() => {
    if (!ref) return
    if (typeof ref === 'function') {
      ref(innerRef.current)
    } else {
      ref.current = innerRef.current
    }
  })

  return innerRef
}
