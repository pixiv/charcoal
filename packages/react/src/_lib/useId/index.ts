import { useId as useReactId } from 'react'

/**
 * Custom useId hook that accepts an optional fallback id
 * Returns the provided id if given, otherwise generates a unique id using React's useId
 */
export function useId(id?: string): string {
  const generatedId = useReactId()
  return id ?? generatedId
}
