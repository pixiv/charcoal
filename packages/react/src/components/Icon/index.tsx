import * as React from 'react'

import type { PixivIcon, Props } from '@charcoal-ui/icons'

export interface OwnProps {
  unsafeNonGuidelineScale?: number
  className?: string
}

export interface IconProps
  extends OwnProps,
    React.PropsWithoutRef<
      Omit<Props, 'class' | 'unsafe-non-guideline-scale' | 'css'>
    > {}

const Icon = React.forwardRef<PixivIcon, IconProps>(function IconInner(
  { name, scale, unsafeNonGuidelineScale, className, ...rest },
  ref
) {
  return (
    <pixiv-icon
      ref={ref}
      name={name}
      scale={scale}
      unsafe-non-guideline-scale={unsafeNonGuidelineScale}
      class={className}
      {...rest}
    />
  )
})

export default Icon
