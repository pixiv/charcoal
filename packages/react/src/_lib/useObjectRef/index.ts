import {
  useCallback,
  useMemo,
  useRef,
  type MutableRefObject,
  type Ref,
} from 'react'
import { mergeRefs } from '../mergeRefs'

/**
 * Custom implementation of useObjectRef to replace @react-aria/utils
 * Converts any ref to a mutable ref object
 */
export function useObjectRef<T>(
  forwardedRef?: Ref<T>
): MutableRefObject<T | null> {
  const objRef = useRef<T | null>(null)
  const _ref = useMemo(() => mergeRefs(objRef, forwardedRef), [forwardedRef])

  const ref = useCallback((value: T | null) => _ref(value), [_ref])

  // Return an object that can be used as both a callback ref and a mutable ref
  return {
    get current() {
      return objRef.current
    },
    set current(value: T | null) {
      ref(value)
    },
  }
}
