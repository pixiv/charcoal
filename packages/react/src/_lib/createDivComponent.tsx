import { DetailedHTMLProps, DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES, forwardRef, ForwardRefExoticComponent, HTMLAttributes, RefAttributes, RefObject } from 'react'
import { useClassNames } from './useClassNames'

export function createDivComponent(mainClassName: string): ForwardRefExoticComponent<Omit<Omit<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "ref"> & {
  ref?: ((instance: HTMLDivElement | null) => void | DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES[keyof DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES]) | RefObject<HTMLDivElement> | null | undefined
}, "ref"> & RefAttributes<HTMLDivElement>> {
  return forwardRef<HTMLDivElement, React.ComponentPropsWithRef<'div'>>(
    function DivComponent({ className, ...props }, ref) {
      const classNames = useClassNames(mainClassName, className)
      return <div className={classNames} ref={ref} {...props} />
    }
  )
}
