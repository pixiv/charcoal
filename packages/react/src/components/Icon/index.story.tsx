import React from 'react'
import Icon, { IconProps } from '.'
import { KNOWN_ICON_FILES } from '@charcoal-ui/icons'
import { Story } from '../../_lib/compat'

export default {
  title: 'Icon',
  component: Icon,
  argTypes: {
    name: {
      control: {
        type: 'select',
        options: KNOWN_ICON_FILES,
      },
    },
    scale: {
      control: {
        type: 'select',
        options: [1, 2, 3],
      },
    },
  },
}

// NOTICE: コード生成のタイミングで不当に型エラーが出ることがあるので name を any にする
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Template: Story<IconProps & { name: any }> = (props) => (
  <Icon {...props} />
)

export const Default = Template.bind({})

Default.args = { name: KNOWN_ICON_FILES[0], scale: 1 }
