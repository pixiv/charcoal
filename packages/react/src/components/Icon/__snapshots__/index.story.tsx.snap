// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Icon Default 1`] = `
<div
  data-dark={false}
>
  <pixiv-icon
    name="16/Add"
    scale={1}
  />
</div>
`;
